### PR TITLE
[ROCm] Enable BFloat16 support for Convolutions on ROCm

### PR DESCRIPTION
**Pull request by rohithkrn**
This PR adds bfloat16 support for convolutions on ROCm.

- Intergrates MIOpen bfloat16 convolution support into PyTorch

- Enables bfloat16 convolution for non-miopen paths, i.e THCUNN, native hip kernels

- Enables bfloat16 type for probability distribution functions(this is included in this PR since conv unit tests use bfloat16 random number generators)

Native cuda kernels for convolution and random functions will be compiled for CUDA as well.

@iotamudelta @bddppq

### CODE REVIEW

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -36,7 +36,7 @@ struct ConvParams {
   bool use_cpu_depthwise3x3_winograd(const at::Tensor& input, const at::Tensor& weight) const;
   bool use_cudnn(const at::Tensor& input) const;
   bool use_cudnn_depthwise(const at::Tensor& input, const at::Tensor& weight) const;
-  bool use_miopen(const at::Tensor& input) const;
+  bool use_miopen(const at::Tensor& input, const bool& bias_defined) const;
```
Comment (ezyang) [inline aten/src/ATen/native/Convolution.cpp]: Don't pass booleans by const reference, it's pointless

**Review by ezyang — APPROVED** [review]
Looks legit. Please find someone not me to land this if you want it landed before the new year.

File: aten/src/ATen/native/Convolution.cpp (hunk 1 above)
Comment (rohithkrn) [reply aten/src/ATen/native/Convolution.cpp]: Agree.

**Comment by iotamudelta** [comment]
ping @bddppq @gottbrath

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 4acc649b:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=4acc649b43021d96d06e0070a1d262216ebd6e25).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4098233) pytorch_ios_11_2_1_x86_64_build (1/1)
**Step:** "Run Simulator Tests" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=79267664) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=79267664))
<details>
<summary>
<code>ImportError: cannot import name 'PILLOW_VERSION' from 'PIL' (/Users/distiller/anaconda/lib/python3.7/site-packages/PIL/__init__.py)</code>
</summary>

```
  File "/Users/distiller/anaconda/lib/python3.7/site-packages/torchvision/datasets/__init__.py", line 9, in <module> 
    from .fakedata import FakeData 
  File "/Users/distiller/anaconda/lib/python3.7/site-packages/torchvision/datasets/fakedata.py", line 3, in <module> 
    from .. import transforms 
  File "/Users/distiller/anaconda/lib/python3.7/site-packages/torchvision/transforms/__init__.py", line 1, in <module> 
    from .transforms import * 
  File "/Users/distiller/anaconda/lib/python3.7/site-packages/torchvision/transforms/transforms.py", line 17, in <module> 
    from . import functional as F 
  File "/Users/distiller/anaconda/lib/python3.7/site-packages/torchvision/transforms/functional.py", line 5, in <module> 
    from PIL import Image, ImageOps, ImageEnhance, PILLOW_VERSION 
ImportError: cannot import name 'PILLOW_VERSION' from 'PIL' (/Users/distiller/anaconda/lib/python3.7/site-packages/PIL/__init__.py) 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Pull request closed without merging**